File: 백준 문제 풀이 모음/class 3/14500.py
# import sys

# n, m = map(int, sys.stdin.readline().split())

# graph = []

# for i in range(n) :
#     tmp = list(map(int, sys.stdin.readline().split()))
#     graph.append(tmp)


# from collections import deque

# def bfs(a, b, cost) :
#     visited = []
#     for i in range(n) :
#         tmp = [False] * m
#         visited.append(tmp)
#     visited[a][b] = True

#     dx = [1, -1, 0, 0]
#     dy = [0, 0, 1, -1]

#     queue = deque()
#     queue.append([(a, b, cost)])

#     count = 0
    
#     while queue :


#         node = queue.popleft()
#         if count == 3 :
#             taa = []
#             for maxmax in node :
#                 taa.append(maxmax[2])
            
#             result =  graph[a][b]

#             return max(taa)
#             break

#         count_list = []
#         for z in range(len(node)) :
#             x, y, costcost = node[z][0], node[z][1], node[z][2]
#             visited[x][y] = True
#             for i in range(4) :
#                 ax = dx[i] + x
#                 ay = dy[i] + y

#                 if ax <0 or ax >= n or ay < 0 or ay >= m :
#                     continue
                
#                 if visited[ax][ay] == True :
#                     continue
                
#                 if visited[ax][ay] == False :
                    
#                     count_list.append((ax, ay, costcost + graph[ax][ay]))
        
#         queue.append(count_list)
#         count += 1


# def block(r, c, total) :
#     make_block = 0
#     dr = [-1, 0, 1, 0]
#     dc = [0, 1, 0, -1]

#     block_list = []
#     for i in range(4) :
#         nr = r +dr[i]
#         nc = c +dc[i]
#         if 0 <= nr < n and 0 <= nc < m :
#             make_block += 1
#             total += graph[nr][nc]
#         if make_block == 3 :
#             block_list.append(total)
#             total -= graph[nr][nc]
#         if make_block == 4:
#             for i in range(4) :
#                 nr = r + dr[i]
#                 nc = c + dc[i]
                
#                 block_list.append(total)

    
#     if len(block_list)== 0 :
#         return 0
#     else :
#         return max(block_list)


# cost = []
# for i in range(n) :
#     for j in range(m) :
#         tmp = bfs(i, j, graph[i][j])
#         total = block(i, j, graph[i][j])
        
#         cost.append(max([tmp, total]))



# print(max(cost))


import sys; input = sys.stdin.readline
def dfs(r, c, idx, total):
    global ans
    if idx == 3:
        if total > ans :
            ans = total
    else :
        for i in range(4) :
            nr = r + dr[i]
            nc = c + dc[i]
            if 0 <= nr < N and 0 <=nc < M :
                if visit[nr][nc] == 0:
                    visit[nr][nc] = 1
                    dfs(nr, nc, idx + 1, total + arr[nr][nc])
                    visit[nr][nc] = 0

def block(r, c, total):
    global ans
    make_block = 0
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        if 0 <= nr < N and 0 <= nc < M:
            make_block += 1
            total += arr[nr][nc]

    if make_block == 3:
        if total > ans:
            ans = total

    if make_block == 4:
        for i in range(4):
            nr = r + dr[i]
            nc = c + dc[i]
            total -= arr[nr][nc]
            if total > ans:
                ans = total
            total += arr[nr][nc]

N, M = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
visit = [([0] * M) for _ in range(N)]
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]
ans = 0

for r in range(N):
    for c in range(M):
        visit[r][c] = 1
        dfs(r, c, 0, arr[r][c])
        block(r, c, arr[r][c])
        visit[r][c] = 0

print(ans)

File: 백준 문제 풀이 모음/kotlin/class 3/14500.kt



var matrix = MutableList<MutableList<Int>>(0) { MutableList<Int>(0) { 0 } }
var visited = MutableList<MutableList<Boolean>>(0) { MutableList<Boolean>(0) { false } }
var count = 0
var result = 0
var dr = listOf<Int>(1,-1,0,0)
var dc = listOf<Int>(0,0,1,-1)


fun main() = with(System.`in`.bufferedReader()) {
    val input = readLine().split(' ').toList()
    val rol = input[0].toInt()
    val col = input[1].toInt()

    matrix = MutableList<MutableList<Int>>(rol) { MutableList<Int>(col) { 0 } }

    for (i in 0 until rol) {
        val str = readLine().split(' ').toList()
        for (j in 0 until col) {
            matrix[i][j] = str[j].toInt()
        }
    }

    visited = MutableList<MutableList<Boolean>>(rol) { MutableList<Boolean>(col) { false } }

    for (i in 0 until rol) {
        for (j in 0 until col) {
            visited[i][j] = true
            var idx = 0
            DFS(idx, i, j, rol, col, 0)
            BLOCK(i, j, rol, col)

            visited[i][j] = false
        }
    }



    println(result)
}

fun DFS(idx: Int, i: Int, j:Int, rol: Int, col: Int, count:Int) {
    if (idx == 4) {
        result = Math.max(result, count)
        return
    }

    for (x in 0 until 4) {
        val new_r = i + dr[x]
        val new_c = j + dc[x]
        if (new_r < 0 || new_r >= rol || new_c < 0 || new_c >= col) {continue}
            if (visited[new_r][new_c] == false) {
                visited[new_r][new_c] = true
                DFS(idx + 1, new_r, new_c, rol, col, count + matrix[i][j])
                visited[new_r][new_c] = false
            }
    }
}

fun BLOCK(i: Int, j: Int, rol: Int, col: Int) {
    val coord = mutableListOf<Pair<Int, Int>>()

    for (x in 0 until 4) {
        val new_r = i + dr[x]
        val new_c = j + dc[x]
        if (new_r < 0 || new_r >= rol || new_c < 0 || new_c >= col) {continue}

        coord.add(Pair(new_r, new_c))
    }


    if (coord.size == 3) {
        result = Math.max(result, matrix[coord[0].first][coord[0].second] + matrix[coord[1].first][coord[1].second] + + matrix[coord[2].first][coord[2].second] + matrix[i][j])
    } else if (coord.size == 4) {
        result = Math.max(result, matrix[coord[0].first][coord[0].second] + matrix[coord[1].first][coord[1].second] + + matrix[coord[2].first][coord[2].second] + matrix[i][j])
        result = Math.max(result, matrix[coord[0].first][coord[0].second] + matrix[coord[1].first][coord[1].second] + + matrix[coord[3].first][coord[3].second] + matrix[i][j])
        result = Math.max(result, matrix[coord[0].first][coord[0].second] + matrix[coord[2].first][coord[2].second] + + matrix[coord[3].first][coord[3].second] + matrix[i][j])
        result = Math.max(result, matrix[coord[1].first][coord[1].second] + matrix[coord[2].first][coord[2].second] + + matrix[coord[3].first][coord[3].second] + matrix[i][j])
    }
}
